### PR TITLE
feat: check commit messages to see if a new version needs to be published

### PR DESCRIPTION
**Pull request by mdonnalley**
### What does this PR do?
Checks commit messages to see if we need to publish a new version or not. For example
  - if only a `chore` commit is found, then it will exit early.
  - if both a `chore` and `feat` commit is found, then it will continue with the release

TODO:
- [x] fix broken unit tests

### What issues does this PR fix or reference?
@W-8775832@
@W-8820671@

### CODE REVIEW

**Review by peternhale — APPROVED** [review]

File: src/repository.ts [hunk 1]
```diff
@@ -190,6 +201,47 @@ abstract class Repository extends AsyncOptionalCreatable {
     return found;
   }
 
+  /**
+   * Analyzes the commits to see if any of them indicate that a new
+   * release should be published.
```
Comment (amphro) [inline src/repository.ts:206]: Can we put in the comment that this is a workaround with a link to the issue on standard-version. I would like to remove this down the road if they update that library.

File: src/repository.ts [hunk 2]
```diff
@@ -390,7 +444,7 @@ export class SinglePackageRepo extends Repository {
   protected async init(): Promise<void> {
     this.logger = await Logger.child(this.constructor.name);
     this.package = await Package.create();
-
+    this.shouldBePublished = await this.isReleasable(this.package);
```
Comment (amphro) [inline src/repository.ts:447]: So here is a question. If we force update the version in `package.json` without a `fit` or `feat` commit message, do we still want to release it? For example, maybe in `salesforcedx` we have a commit message on Thursday as `chore: release 50.17.0` where we hardcode the version in that commit. In that case we want to release. It feels weird to have the commit be `feat: release 50.17.0`.

Thoughts?

**Review by amphro — CHANGES_REQUESTED** [review]

File: src/repository.ts (hunk 2 above)
Comment (amphro) [reply src/repository.ts:447]: Maybe to stay consistent we do use `feat` to kick off the release... just pointing it out.

**Comment by amphro** [comment]
Can we also add some information to a readme somewhere if it isn't already? Seems like it should be doc'ed

File: src/repository.ts (hunk 2 above)
Comment (mdonnalley) [reply src/repository.ts:447]: That's a good point. I'll update it to ignore this check if the version bump has been hardcoded in the package.json

File: README.md [hunk 3]
```diff
@@ -1,8 +1,34 @@
...
+
+The following steps are automated for package releases
+
+### Version Bump
+
+We use [standard-version](https://github.com/conventional-changelog/standard-version) to determine the next version that will be published. This means that all commits **must** adhere to the [conventional commits specification](https://www.conventionalcommits.org/en/v1.0.0/) in order for `standard-version` to work.
+
+In the case the you have manually bumped the version in the package.json, then the plugin will respect that and publish that version instead of using `standard-version` to determine the next version.
+
+NOTE: We consider the `chore`, `style`, `docs`, `ci`, `test` commit types to be "non-releasable", meaning that if all the commits are of those types then we do not publish a new version. However, if you've manually bumped the version in the package.json then the plugin will publish that version regardless of the commit types.
```
Comment (amphro) [inline README.md:15]: Sweet. Thanks.

File: src/repository.ts [hunk 4]
```diff
@@ -21,6 +23,9 @@ import { upload } from './codeSigning/upload';
 import { Package, VersionValidation } from './package';
 import { Registry } from './registry';
 
+const conventionalCommitsParser = require('conventional-commits-parser');
```
Comment (amphro) [inline src/repository.ts:26]: Why do these need to be brought in as require? Can't you still do `import * as conventionalCommitsParser from 'conventional-commits-parser'`?

File: src/repository.ts [hunk 5]
```diff
@@ -192,14 +203,63 @@ abstract class Repository extends AsyncOptionalCreatable {
     return found;
   }
 
+  /**
+   * If the commit type isn't fix (patch bump), feat (minor bump), or breaking (major bump),
+   * then standard-version always defaults to a patch bump.
+   * See https://github.com/conventional-changelog/standard-version/issues/577
```
Comment (amphro) [inline src/repository.ts:209]: Thank you.

**Review by amphro — APPROVED** [review]

File: src/repository.ts [hunk 6]
```diff
@@ -288,8 +348,9 @@ export class LernaRepo extends Repository {
     if (!isEmpty(nextVersions)) {
       for (const pkgPath of pkgPaths) {
         const pkg = await Package.create(pkgPath);
+        const shouldBePublihsed = await this.isReleasable(pkg, true);
         const nextVersion = get(nextVersions, `${pkg.name}.nextVersion`, null) as string;
```
Comment (shetzel) [inline src/repository.ts:352]: nit: `get()` --> `getString()`

**Review by shetzel — APPROVED** [review]
just some nits, but approved

**Comment by amphro** [comment]
*Testing notes:*

If I have *no type* or a *random invalid type*, it will still publish. You could argue that is a problem with the repo not validating commit messages but it could also be bypassed by the command line or UI. In that case, maybe it should be published. If the answer is on, Instead of `skippableCommitTypes` we should have `commitTypes: ['feat', 'feat!', 'bug', 'bug!', 'perf', 'refactor']`.  I'm fine keeping it skippable too. Either way, it is a list we need to maintain.

I published ~15 package versions with a variety of commit messages in between each release testing patch, minor, and major. Other than the issue above it works great. 

I'll wait for @mdonnalley feedback on the issue before merging.

**Comment by mdonnalley** [comment]
@amphro I think it's fine to allow it to publish in those scenarios. If someone has gone through the trouble of bypassing the commit linting, then they are probably (hopefully) doing it purposely. That being said, we can always address that scenario later if it becomes an issue

**Comment by amphro** [comment]
Sounds good to me. I just wanted to make sure we had the conversation.

**Pull request merged**